### PR TITLE
Add transitive dependencies flag for yarn audit

### LINKED ISSUE #997: New dependency check step for yarn v3 does not capture devDependencies
### What would you like to change?
Add the -R flag to the `yarn npm audit` lines within `yarnV2OrNewer-audit-with-suppressions.sh`. This will restore it to the behaviour of the old yarn audit from v1.

### How do you think that would improve the project?
Beefs up security to what it was pre-v3 upgrade.

### If this entry is related to a bug, please provide the steps to reproduce it
- Open a repo with vulnerabilities in peer dependencies
- Use `yarn set version classic` in your terminal emulator of choice
- Run `yarn audit` and record the vulnerabilities
- Use `yarn set version berry` to switch back to v3
- Run `yarn npm audit` and record the vulnerabilities (or lack thereof)
- Run `yarn npm audit -R` and record the vulnerabilities

### PR DESCRIPTION
**Pull request by furnivall**
Resolves #997 . (This is applicable only if this pull request relates to a GitHub issue, delete the line otherwise)

Notes:
* Should capture all dependencies ala old-style yarn audit

### CODE REVIEW

**Comment by furnivall** [comment]
It's worth mentioning that the "failed tests" on this build actually highlight the issue perfectly! https://sandbox-build.platform.hmcts.net/blue/organizations/jenkins/Pipeline_Test_Sandbox%2Fcnp-plum-frontend/detail/master/965/pipeline

File: resources/uk/gov/hmcts/pipeline/yarn/yarnV2OrNewer-audit-with-suppressions.sh
Comment (timja) [inline resources/uk/gov/hmcts/pipeline/yarn/yarnV2OrNewer-audit-with-suppressions.sh:6]:
```suggestion
yarn npm audit --recursive --environment production
```

File: resources/uk/gov/hmcts/pipeline/yarn/yarnV2OrNewer-audit-with-suppressions.sh
Comment (timja) [inline resources/uk/gov/hmcts/pipeline/yarn/yarnV2OrNewer-audit-with-suppressions.sh:9]:
```suggestion
yarn npm audit --recursive --environment production --json > yarn-audit-result
```

**Review by timja — CHANGES_REQUESTED** [review]
This seems okay although we may want to wait a bit before introducing it to let things settle for teams as they have a couple of technical workpieces to complete currently and it will need announcing before merging.

----

On the implementation please don't use short flags in scripts, long flags are self-descriptive, short flags need someone to look them up in the manual which makes the scripts harder to read.

----

Have you tested this on a couple of projects to see the impact of this change?

**Comment by furnivall** [comment]
> This seems okay although we may want to wait a bit before introducing it to let things settle for teams as they have a couple of technical workpieces to complete currently and it will need announcing before merging.
> 
> On the implementation please don't use short flags in scripts, long flags are self-descriptive, short flags need someone to look them up in the manual which makes the scripts harder to read.
> 
> Have you tested this on a couple of projects to see the impact of this change?

Just realised I never responded to this, although I did merge your changes. I think it's ready to go whenever you're ready to push the button. It's been tested on P&I test branches and the pipeline itself highlights the issue on this PR.

**Comment by timja** [comment]
Possible for you to draft a pre announcement? e.g. on x day (at least 1-2 weeks) we'll be merging this change, this affects frontend repositories, run this ahead of time to fix it so you won't be blocked on the day

Thanks

**Comment by timja** [comment]
(and to get a green build here)

**Comment by furnivall** [comment]
> Possible for you to draft a pre announcement? e.g. on x day (at least 1-2 weeks) we'll be merging this change, this affects frontend repositories, run this ahead of time to fix it so you won't be blocked on the day
> 
> Thanks

Absolutely, will draft something now.

File: src/uk/gov/hmcts/contino/YarnBuilder.groovy
Comment (timja) [inline src/uk/gov/hmcts/contino/YarnBuilder.groovy:162]: hmm I'm not sure if this is quite right, a nagger should only alert if it detects there's something wrong. which in this case would be to run both and only trigger the nagger if recursive fails but the current version passes.

(which would depend on how much time it adds to the build to add both, if its small then fine, if its not small then an announcement on slack is probably better)

Comment (furnivall) [reply src/uk/gov/hmcts/contino/YarnBuilder.groovy]: Should now work as requested :)

File: src/uk/gov/hmcts/contino/YarnBuilder.groovy
Comment (timja) [inline src/uk/gov/hmcts/contino/YarnBuilder.groovy:318]: please indent properly.
if block close } and function } shouldn't be at the same level

File: src/uk/gov/hmcts/contino/YarnBuilder.groovy
Comment (timja) [inline src/uk/gov/hmcts/contino/YarnBuilder.groovy:317]:
```suggestion
        "We will be adding the --recursive flag to CVE scanning to check your transitive dependencies, your security scan is currently failing with these dependencies added to the scan. Run `yarn npm audit --recursive --environment production` ahead of time to catch and fix issues so you won't be blocked on the day of the change.", LocalDate.of(2023, 03, 15))
```

**Review by timja — APPROVED** [review]

File: resources/uk/gov/hmcts/pipeline/yarn/yarnV2OrNewer-audit-with-suppressions.sh
Comment (timja) [inline resources/uk/gov/hmcts/pipeline/yarn/yarnV2OrNewer-audit-with-suppressions.sh:7]: date needs updating

File: src/uk/gov/hmcts/contino/YarnBuilder.groovy
Comment (timja) [inline src/uk/gov/hmcts/contino/YarnBuilder.groovy:320]: date needs updating

**Review by timja — APPROVED** [review]

**Review by timja — APPROVED** [review]

**Review by timja — APPROVED** [review]

**Pull request merged**
